lib/logstash/outputs/s3.rb

```diff
@@ -397,16 +398,43 @@ def restore_from_crash
     @crash_uploader = Uploader.new(bucket_resource, @logger, CRASH_RECOVERY_THREADPOOL)
 
     temp_folder_path = Pathname.new(@temporary_directory)
-    Dir.glob(::File.join(@temporary_directory, "**/*"))
-      .select { |file| ::File.file?(file) }
-      .each do |file|
-      temp_file = TemporaryFile.create_from_existing_file(file, temp_folder_path)
-      if temp_file.size > 0
-        @logger.debug? && @logger.debug("Recovering from crash and uploading", :path => temp_file.path)
-        @crash_uploader.upload_async(temp_file, :on_complete => method(:clean_temporary_file), :upload_options => upload_options)
-      else
-        clean_temporary_file(temp_file)
+
+    files = Dir.glob(::File.join(@temporary_directory, "**/*"))
+               .select { |file_path| ::File.file?(file_path) }
+    under_recovery_files = get_under_recovery_files(files)
+
+    files.each do |file_path|
+      # if already recovering or recovered and uploading to S3, skip the files
+      unless under_recovery_files.include?(file_path)
+        temp_file = TemporaryFile.create_from_existing_file(file_path, temp_folder_path)
+        if temp_file.size > 0
+          @logger.debug? && @logger.debug("Recovering from crash and uploading", :path => temp_file.path)
+          @crash_uploader.upload_async(temp_file,
+                                       :on_complete => method(:clean_temporary_file),
+                                       :upload_options => upload_options)
+        end
+        if @encoding != GZIP_ENCODING && temp_file.size != 0
+          clean_temporary_file(temp_file)
+        end
+      end
+    end
+  end
+
+  # figures out the recovering files and
+  # creates a skip list to ignore for the rest of processes
+  def get_under_recovery_files(files)
+    skip_files = Set.new
+    return skip_files unless @encoding == GZIP_ENCODING
+
+    files.each do |file_path|
+      if file_path.include?(TemporaryFile.recovery_file_name_tag)
+        skip_files << file_path
+        if file_path.include?(TemporaryFile.gzip_extension)
+          # also include the original corrupted gzip file
+          skip_files << file_path.gsub(TemporaryFile.recovery_file_name_tag, "")
+        end
       end
     end
+    skip_files
   end
```

Comment:
Somewhat of an edge case, but if we skip the original zip file, if a failure/crash occurs during a valid recovery, the directory will be in a state of 'limbo' where the contents of the folder will be neither deleted nor uploaded.

Not including the original zip file in this list would help recovery in this case, but will likely require some additional logic to handle unrecoverable errors (maybe an indication that the file is "damaged"), and remove recovery that is half finished?

```suggestion
  def get_under_recovery_files(files)
    files.select { |file_path| file_path.include?(TemporaryFile.recovery_file_name_tag)}
 ```

Reply:
The only possible case I can see is when S3 is unavailable (or LS is having an issue with S3). We are not removing the temporary folder unless we achieve the successful completion of recovery. And I think it is better to get know the cases better first.

Also, suggested logic will not work because as you can see L431 & L434 lines are adding two entries where suggestion will add only one.

Reply:
Yes, that was the scenario that I was testing out. My main concern with the current approach is that we are skipping over files in the temporary directory without informing the user - I think we should either log that we are skipping over those files, or attempt to re-upload the zip file (logging that too)

(my suggested code snippet was to include the original zip file to attempt to re-upload it, rather than a like for like swap of your original code)

Reply:
I'm ok with logging - I'd much rather that than silently skipping over. We can iterate over this in a future PR